README.md
# django-maintenancemode-2

[![Build Status](https://travis-ci.org/alsoicode/django-maintenancemode-2.svg)](https://travis-ci.org/alsoicode/django-maintenancemode-2)

Current Version: 2.0.0

This project makes it easy to put your Django site into "maintenance mode", or more technically, return an HTTP 503 response.

This project differs slightly from other implementations in that the maintenance mode flag is stored in your database versus settings or an environment variable. If your site is deployed to multiple servers, the centralized database-based maintenance flag makes it a snap to bring them all up or down at once.

## Requirements
- [django](https://www.djangoproject.com/download/)
- [django.contrib.sites](https://docs.djangoproject.com/en/1.11/ref/contrib/sites/)

## Pre-Requisites
You must have at least one Site entry in your database **before** installing django-maintenancemode-2.

## Supported Python Versions
- 2.7, 3.x

## Supported Django Versions
- 4.x use the latest version
- 2.x >= 3.x, please use version 1.3.1
- < 2, please use version 1.1.9

## Installation
1. `pip install django-maintenancemode-2`

-- or --

1. Download django-maintenancemode-2 from [source](https://github.com/alsoicode/django-maintenancemode-2/archive/master.zip)
2. *optional: Enable a virtualenv
3. Run `python setup.py install` or add `maintenancemode` to your PYTHONPATH

## Settings and Required Values
- Ensure the [Sites Framework](https://docs.djangoproject.com/en/1.11/ref/contrib/sites/) is enabled, and you have at least one entry in the Sites table.
- Add `maintenancemode.middleware.MaintenanceModeMiddleware` to your `MIDDLEWARE_CLASSES`
- Add `maintenancemode` to your `INSTALLED_APPS`
- Run `python manage.py migrate` to create the `maintenancemode` tables.
- Run your project to automatically add the `maintenancemode` database records.
- Add a 503.html template to the root of your templates directory, or optionally add a `MAINTENANCE_503_TEMPLATE` path to your 503.html file's location in settings.
- `maintenancemode` will ignore any patterns beginning with the default Django Admin url: `^admin` so you can turn it off. If you use a custom url for admin, you may override the ignored admin patterns by adding the `MAINTENANCE_ADMIN_IGNORED_URLS` list in settings. Example: `['^my-custom-admin', '^my-other-custom-admin']`
- You can also block staff users, who by default are ignored by maintenance mode, by setting `MAINTENANCE_BLOCK_STAFF` to `True`

## Usage

![Image of django-maintenancemode-2](http://res.cloudinary.com/alsoicode/image/upload/v1449537052/django-maintenancemode-2/maintenancemode.jpg)

### Turning Maintenance Mode **On**
To put a site into "Maintenance Mode", check the "In Maintenance Mode" checkbox and save in Django Admin under the "Maintenancemode" section. The next time you visit the public side of the site, it will return a 503 if:

- You are not logged in as a superuser or staff user
- You are not viewing a URL in the ignored patterns list
- Your `REMOTE_ADDR` does not appear in the `INTERNAL_IPS` setting


Or you can alternatively use the `setmaintenance` management command:

```
    # sets maintenance on for the current settings.SITE_ID
    ./manage.py setmaintenance on

    # sets maintenance on for sites 2 and 3
    ./manage.py setmaintenance on 2 3
```

which can be useful for `fabric` deployment scripts etc.

### Turning Maintenance Mode **Off**
Log in, un-check the "In Maintenance Mode" checkbox and save.

Or you can alternatively use the `setmaintenance` management command:

```
    # sets maintenance off for the current settings.SITE_ID
    $ ./manage.py setmaintenance off

    # sets maintenance off for sites 2 and 3
    $ ./manage.py setmaintenance off 2 3
```

## Testing and Sample Application
A "testproject" application is included, which also contains unit and functional tests you can run via `python manage.py test` from the `testproject` directory.

You will need to run `manage.py migrate` to create the test project database.

There are only two views in the testproject:
- /
- /ignored-page

To see `maintenancemode` in action, log into Django admin, and set the maintenance mode to true. Log out, then visit the home page, and instead, you'll be greeted with the maintenance page.

To have `maintenancemode` ignore the "ignored-page" view, simply add it's url pattern to the Ignored URLs as:

    ^ignored-page/$

Now you should be able to visit the `ignored-page` view regardless of the maintenancemode status. This is useful for contact or help pages you still want people to be able to access while you're working on other parts of the site.

### Database migrations
`./manage.py migrate` should add the necessary tables.


CHANGES
Changes
== == == =

1.1.8
- - - - -
- Added Python 3.6 support.
- Added Django 1.11 support.
- Documentation updates.
- Syntax improvements.

1.1.7
- - - - -
Added `setmaintenance` management command.

1.1.6
- - - - -
Added Django 1.10 support.

1.1.5
- - - - -
Simplified admin ignore pattern settings.

1.1.4
- - - - -
Fixed logging into Django admin when maintenance mode is enabled.

1.1.3
- - - - -
Converstion of README.md to .rst for better compatibility with PyPi

1.1.2
- - - - -
- Documentation updates.
- Added setting for 503.html template path

1.1.1
- - - - -
- Deleted

1.1.0
- - - - -
- Updated testproject
- Django 1.8 compatibility.

0.9.3
- - - - -
- Moved maintenance mode and ignored url patterns from settings.py to database backed storage.


0.9.2
- - - - -
- Fixed an issue with setuptools, thanks for reporting this ksato9700

0.9.1
- - - - -

- Tested django - maintenancemode with django - 1.0 release (following the 1.0.X release branch)
- Bundled buildout.cfg and bootstrap with the source version of the project, allowing repeatable buildout
- The middleware now uses its own default config file, thanks to a patch by semente
- Use INTERNAL_IPS to check for users that need access. user.is_staff will stay in place for backwards incompatibility. Thanks for the idea Joshua Works
- Have setup.py sdist only distribute maintenancemode itself, no longer distribute tests and buildout stuff
- Use README and CHANGES in setup.py's long_description, stolen from Jeroen's djangorecipe :)
- Updated the documentation and now use pypi as the documentation source (link there from google code)

0.9
- - - - -

First release


testproject/testproject/settings.py
"""
Django settings for testproject project.

Generated by 'django-admin startproject' using Django 1.8.7.

For more information on this file, see
https://docs.djangoproject.com/en/1.8/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/1.8/ref/settings/
"""

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
import django
import os

DJANGO_MAJOR_VERSION = django.VERSION[0]
DJANGO_MINOR_VERSION = django.VERSION[1]

# Optionally, specify a custom 503 template path
# MAINTENANCE_503_TEMPLATE = 'errors/503.html'

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/1.8/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = 'k=t$%@*+@=5c57g&f^&8$)evztgb1%b4l23zt*!2e-1))3@vue'

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = True

SITE_ID = 1

ALLOWED_HOSTS = []


# Application definition

INSTALLED_APPS = (
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.messages',
    'django.contrib.sessions',
    'django.contrib.sites',
    'django.contrib.staticfiles',

    'app',
    'maintenancemode',
)


if DJANGO_MAJOR_VERSION == 1 and DJANGO_MAJOR_VERSION < 10:
    MIDDLEWARE_CLASSES = (
        'django.contrib.sessions.middleware.SessionMiddleware',
        'django.middleware.common.CommonMiddleware',
        'django.middleware.csrf.CsrfViewMiddleware',
        'django.contrib.auth.middleware.AuthenticationMiddleware',
        'django.contrib.messages.middleware.MessageMiddleware',
        'django.middleware.clickjacking.XFrameOptionsMiddleware',
    )
else:
    MIDDLEWARE = [
        'django.middleware.security.SecurityMiddleware',
        'django.contrib.sessions.middleware.SessionMiddleware',
        'django.middleware.common.CommonMiddleware',
        'django.middleware.csrf.CsrfViewMiddleware',
        'django.contrib.auth.middleware.AuthenticationMiddleware',
        'django.contrib.messages.middleware.MessageMiddleware',
        'django.middleware.clickjacking.XFrameOptionsMiddleware',
    ]

TEMPLATE_DIRECTORY = os.path.join(BASE_DIR, 'templates')

if DJANGO_MAJOR_VERSION >= 1:

    # Templates
    if DJANGO_MINOR_VERSION < 8 and DJANGO_MAJOR_VERSION == 1:
        TEMPLATE_DIRS = (
            TEMPLATE_DIRECTORY,
        )

        TEMPLATE_LOADERS = (
            'django.template.loaders.filesystem.Loader',
            'django.template.loaders.app_directories.Loader',
        )

        TEMPLATE_CONTEXT_PROCESSORS = (
            'django.contrib.auth.context_processors.auth',
            'django.core.context_processors.i18n',
            'django.core.context_processors.request',
            'django.core.context_processors.media',
            'django.core.context_processors.static',
            'django.contrib.messages.context_processors.messages',
        )
    else:
        TEMPLATES = [
            {
                'BACKEND': 'django.template.backends.django.DjangoTemplates',
                'DIRS': [TEMPLATE_DIRECTORY],
                'APP_DIRS': True,
                'OPTIONS': {
                    'context_processors': [
                        'django.template.context_processors.debug',
                        'django.template.context_processors.request',
                        'django.contrib.auth.context_processors.auth',
                        'django.contrib.messages.context_processors.messages',
                    ],
                },
            },
        ]

    # Sessions
    if DJANGO_MAJOR_VERSION == 1 and DJANGO_MINOR_VERSION == 7:
        MIDDLEWARE_CLASSES += (
            'django.contrib.auth.middleware.SessionAuthenticationMiddleware',
        )
    elif DJANGO_MAJOR_VERSION == 1:
        MIDDLEWARE_CLASSES += (
            'django.middleware.security.SecurityMiddleware',
        )
    
if DJANGO_MAJOR_VERSION == 1 and DJANGO_MAJOR_VERSION < 10:
    MIDDLEWARE_CLASSES += ('maintenancemode.middleware.MaintenanceModeMiddleware',)
else:
    MIDDLEWARE += ['maintenancemode.middleware.MaintenanceModeMiddleware']

ROOT_URLCONF = 'testproject.urls'

WSGI_APPLICATION = 'testproject.wsgi.application'


# Database
# https://docs.djangoproject.com/en/1.8/ref/settings/#databases

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
    }
}


# Internationalization
# https://docs.djangoproject.com/en/1.8/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_L10N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/1.8/howto/static-files/

STATIC_URL = '/static/'
